### PR TITLE
Plugin: allow cluster chooser button overriding

### PR DESCRIPTION
**Pull request by ashu8912**
# [Title: describe the change in one sentence]
we allow the flexibility to a plugin developer to change the chooser component and replace it with something they want to add.

## Testing done
Added an example plugin that has been tested with the current PR headlamp-plugin types build

### CODE REVIEW

File: plugins/examples/clusterbuttonoverride/src/index.tsx [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+// import { K8s } from '@kinvolk/headlamp-plugin/lib/K8s';
+// import { Typography } from '@material-ui/core';
+
+class ClusterChooserButtonOverridePlugin extends Plugin {
+  initialize(registry: Registry) {
+    console.log('clusterbuttonoverride initialized');
+
+    // Register your plugin methods here, like:
+    // registry.registerSidebarItem(...);
+    // registry.registerRoute(...);
```
Comment (joaquimrocha) [inline plugins/examples/clusterbuttonoverride/src/index.tsx:13]: Remove these and other comments, as this is an example plugin for a particular purpose, not a generic one.

File: plugins/examples/clusterbuttonoverride/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# cra-template-headlamp
+
+This is the official TypeScript template for [Headlamp Plugins](https://github.com/kinvolk/headlamp).
+
+To use this template to start a new plugin with the npx command (comes with Node.js):
+
+```sh
+npx create-react-app headlamp-myfancy --template headlamp
+```
+
```
Comment (joaquimrocha) [inline plugins/examples/clusterbuttonoverride/README.md:10]: This should be updated.

**Review by joaquimrocha — CHANGES_REQUESTED** [review]
Left a few comments to improve things, but overall I think it's definitely in the right direction.

File: plugins/examples/clusterbuttonoverride/src/index.tsx (hunk 1 above)
Comment (ashu8912) [reply plugins/examples/clusterbuttonoverride/src/index.tsx]: done

File: frontend/src/redux/reducers/ui.tsx [hunk 3]
```diff
@@ -123,6 +125,7 @@ export const INITIAL_STATE: UIState = {
   branding: {
     logo: null,
   },
+  clusterChooserButtonComponent: null,
```
[the remark this answers is not in the document]
Comment (ashu8912) [reply frontend/src/redux/reducers/ui.tsx]: updated

File: frontend/src/plugin/registry.tsx [hunk 4]
```diff
@@ -148,4 +151,16 @@ export default class Registry {
   registerAppLogo(component: BrandingProps['logo']) {
     store.dispatch(setBrandingAppLogoComponent(component));
   }
+
+  /**
+   * @param component is a React Component that takes one required props ```JSX clickHandler``` which is the
+   * action handler that happens on the chooser button click
+   * @example
+   * ```JSX
+   * register.registerClusterChooserComponent((props: { clickHandler: () => {} }) => <MY_CUSTOM_COMPONENT onClick={clickHandler}/>)
```
[the remark this answers is not in the document]
Comment (ashu8912) [reply frontend/src/plugin/registry.tsx]: fixed

File: frontend/src/plugin/registry.tsx [hunk 5]
```diff
@@ -16,9 +18,10 @@ export interface SectionFuncProps {
   title: string;
   component: (props: { resource: any }) => JSX.Element | null;
 }
-
+export type clusterChooserButtonComponent = React.ComponentType<{
+  clickHandler: (event?: any) => void;
+}>;
 export type sectionFunc = (resource: KubeObject) => SectionFuncProps | null | undefined;
-
```
[the remark this answers is not in the document]
Comment (ashu8912) [reply frontend/src/plugin/registry.tsx]: fixed

**Comment by illume** [comment]
I think if you rebase this against current main the examples test should pass.

**Review by illume — APPROVED** [review]
👍 Thanks. 🎉🎈

**Pull request merged**